### PR TITLE
add `order_by` param in Scroll API

### PR DESCRIPTION
**Pull request by kwkr**
Hi @generall !
As an exercise to get to know the codebase better, I started implementing #1763 . I was able to hack together some working solution and wanted to confirm whether this is still a required feature and whether I got the requirements right.

So basically I added the `order_by` parameter to the scroll endpoint which allows scrolling through data by using a composite key as pointed out here:
```
Qdrant will read the index sorted by payload key, and will return results sorted by combination of (paylaod_value, id).
```

Currently, there are two possible ways to scroll through the data:
- order_by asc and min param, which will return the data sorted by payload in asc order and then by id.
- order_by desc and max param, which will return the data sorted by payload in desc order and then by id.

After getting the first results, the data can be further scrolled using the combination of payload_value and id.
When no offset is specified, the data is just returned starting from an item according to the provided min/max value.
The solution utilizes the Btrees of the underlying payload indexes so it should be efficient enough.

In both cases, the `filter` param is ignored.

Please let me know if this is still relevant so I can either polish the solution or drop it.

/claim #1763

### CODE REVIEW

**Comment by generall** [comment]
Hi @kwkr, thanks for the PR!


Overall, there are 2 types of concerns I have regarding this task in general.

First, there is an interface concern. Once some fields are introduce, it is very hard for us to change or delete them. That creates a question: how universal the given approach is?
As I can see in your PR, you define min/max values to concrete types, assuming that extension to the other, potentially orderable types, is not expected:

```
    pub min: Option<f64>,
    pub max: Option<f64>,
```

I think a more future-prove option would be to not use any max/min value at all, assuming that we can go to the next page by specifying filter (and this option still leaves the possibility to introduce min/max later).

The second concern is the performance of the solution. Ideally, we would like to use payload index in combination with offset by ID. So, could you please elaborate on the part with

```
while !found_offset {
```

In what cases we would need multiple internal `scroll_by` queries and isn't there somewhere a potential linear scan can happen?

**Comment by kwkr** [comment]
Thanks for the response @generall.
At first, I just followed the requirements from the issue linked above in order to get a working POC. From what I see, the functionality might be also possible to implement using a filter field with range condition (but this one is also f64 only, which would be sufficient for the int/float payloads. The good thing is it doesn't require any new types but the way to actually use it would be pretty convoluted.

There would be also a possibility to introduce the field as some kind of experimental part of API, just to polish the solution in cooperation with the users but I don't know what is your approach towards this type of changes.

When it comes to the second issue:
The current solution uses the payload indexes in order to retrieve the data from the indexes by making range queries to the underlying BTrees. A starting item for a page can be determined by some payload value or by a combination of payload + offset (id). Some preset amount of data is fetched (let's say 100 records) and then the ids for the records are fetched. When the payloads are the same, the item that we are looking for might be absent in the current batch.  For example, we fetched 100 items with price 99.99 but our database contains 150 such items and just by the way the items are organized in the BTree, the item with the offset that we are looking after is in the next batch of items. In order to do our best to find the item with the offset, the queries are repeated until the offset value is found. In the worst case, it can lead to fetching a whole BTree, so I would suggest to set a limit on this (let's say at most 1000 items lookup) and maybe give the user a possibility to adjust this by some optional param.

So to sum it up in one sentence: If we have many items with the same payload, the next criterion is the offset value (id) and we fetch the data from the index until we find the target offset. 

If you are okay with using the filter approach, I can implement it using elements of the current API. My only concern is if this won't interfere with some already existing functionality provided by filters. Also if the logic that I described is not clear, please let me know.

**Comment by kwkr** [comment]
Thanks for the update @coszio ! Do you mean the bounty will still be awarded?

**Comment by coszio** [comment]
Yes 😁

**Comment by coszio** [comment]
I think I've done enough testing now. How about a review round?

**Comment by coszio** [comment]
As a quick recap to what this PR does:
- interface for order-by is a new parameter in scroll:
  ```json
  "order_by": "field_name",

  // or

  "order_by": { "key": "field_name", "direction": "desc" } // or "asc"
  ```
- Only works on keys with numeric indexes
- Order is in (key, id) priority
- Offset still works just with ID because we first fetch this record to get the value offset
- At segment level, a range filter is introduced to only fetch the amount of records needed for returning a page of ordered results
  - start bound is the offset point value
  - end bound is calculated using the numeric index histogram, considering enough records to cover for the cardinality estimation of the user filter.

File: lib/segment/src/segment.rs [hunk 1]
```diff
@@ -1078,6 +1112,109 @@ impl SegmentEntry for Segment {
...
+                filter.map_or(limit, |filter| {
+                    let payload_index = self.payload_index.borrow();
+                    let query_cardinality = payload_index.estimate_cardinality(filter);
+
+                    let available_points = self.available_point_count();
+
+                    let filtered_out_points =
+                        available_points.saturating_sub(query_cardinality.min);
+
+                    // Protect from worst case: all filtered points are in front of the offset
```
Comment (generall) [inline lib/segment/src/segment.rs:1132]: Why is it bad?

File: lib/segment/src/segment.rs [hunk 2]
```diff
@@ -1078,6 +1112,109 @@ impl SegmentEntry for Segment {
...
+            crate::index::field_index::FieldIndex::FloatIndex(float_index) => {
+                let to = float_index.get_range_by_size_excluding(limit, search_from, direction);
+                (from, to)
+            }
+            crate::index::field_index::FieldIndex::IntMapIndex(_)
+            | crate::index::field_index::FieldIndex::KeywordIndex(_)
+            | crate::index::field_index::FieldIndex::GeoIndex(_)
+            | crate::index::field_index::FieldIndex::FullTextIndex(_)
+            | crate::index::field_index::FieldIndex::BinaryIndex(_) => {
+                unimplemented!(
```
Comment (generall) [inline lib/segment/src/segment.rs:1181]: this is a panic. We probably want to return 400 instead.

File: lib/segment/src/types.rs [hunk 3]
```diff
@@ -1045,6 +1045,16 @@ pub enum PayloadFieldSchema {
     FieldParams(PayloadSchemaParams),
 }
 
+impl PayloadFieldSchema {
+    pub fn is_numeric(&self) -> bool {
+        matches!(
+            self,
+            PayloadFieldSchema::FieldType(PayloadSchemaType::Integer)
```
Comment (generall) [inline lib/segment/src/types.rs:1052]: We probably want an exhaustive match here, otherwise it would be very easy to miss this spot in case we invent another numeric index type

File: lib/segment/src/index/field_index/histogram.rs [hunk 4]
```diff
@@ -179,6 +176,79 @@ impl<T: Numericable> Histogram<T> {
         Unbounded
     }
 
+    /// Infers boundaries for bucket of given size and starting point, skipping the first
+    /// border next to the starting point.
+    ///
+    /// E.g:
+    ///                                starts counting in this border
+    ///                                    ---> v
```
Comment (generall) [inline lib/segment/src/index/field_index/histogram.rs:184]: What's? `v`?

File: lib/segment/src/index/field_index/histogram.rs [hunk 5]
```diff
@@ -179,6 +176,79 @@ impl<T: Numericable> Histogram<T> {
...
+    /// 1, 2, 3, 4, 5, 5, 5, 5, 6, 7, 8, 9, 10, 11, 12, 13
+    /// b           [                    b      b       b
```
Comment (generall) [inline lib/segment/src/index/field_index/histogram.rs:186]: What's `b`?

File: lib/segment/src/index/field_index/histogram.rs [hunk 6]
```diff
@@ -179,6 +176,79 @@ impl<T: Numericable> Histogram<T> {
...
+    ///           from
+    pub fn get_range_by_size_excluding(&self, from: Bound<T>, range_size: usize) -> Bound<T> {
```
Comment (generall) [inline lib/segment/src/index/field_index/histogram.rs:188]: Is there a specific reason why we need to involve `histogram` into range reading?

AFAIK, it should be enough to use `range` of a btree (which is a DoubleEndedIterator). I would prefer to avoid using histogram for anything apart from it's original purpose

**Review by generall — CHANGES_REQUESTED** [review]
I agree with interface, but there are questions to the implementation (overall, quite hard to follow), and some design choices.

Also, I would like a bit more detailed comment on what happens if:

- offset is removed from the storage in between scroll calls
- there are multiple numeric values per point

File: lib/collection/src/collection/point_ops.rs [hunk 7]
```diff
@@ -154,6 +155,67 @@ impl Collection {
...
+                };
+
+                // Extract number value from payload
+                offset_point
+                    .iter()
+                    .flatten()
+                    .next()
+                    .and_then(|offset_point| offset_point.payload.as_ref())
+                    .map(|payload| payload.get_value(&order_by.key))
+                    .and_then(|multi_value| multi_value.values().iter().find_map(|v| v.as_f64()))
```
[the remark this answers is not in the document]
Comment (coszio) [reply lib/collection/src/collection/point_ops.rs:213]: Yes, but only if they are in different pages. I considered it to not be expected to work correctly. Otherwise we could only consider first value of multi-value to make paging always work

File: lib/segment/src/index/field_index/histogram.rs (hunk 4 above)
Comment (coszio) [reply lib/segment/src/index/field_index/histogram.rs:184]: Sorry, tried to make a down arrow ⬇️

File: lib/segment/src/index/field_index/histogram.rs (hunk 6 above)
Comment (coszio) [reply lib/segment/src/index/field_index/histogram.rs:188]: 💡 I actually didn't consider to use the `Iterator` part of the `range` function as a streaming source. It's a good idea, let me try to do that instead.

File: lib/segment/src/index/field_index/histogram.rs (hunk 4 above)
Comment (timvisee) [reply lib/segment/src/index/field_index/histogram.rs:184]: ↓

File: lib/segment/src/index/field_index/histogram.rs (hunk 5 above)
Comment (timvisee) [reply lib/segment/src/index/field_index/histogram.rs:186]: I haven't looked into the details yet, but maybe you can denote it with `|`.

File: lib/segment/src/index/field_index/histogram.rs [hunk 8]
```diff
@@ -267,7 +337,13 @@ impl<T: Numericable> Histogram<T> {
                     } else {
                         0
                     };
-                    let max_estimate = a_count.right + 1;
+                    let mut max_estimate = a_count.right + 1;
+
+                    if b_count.right == 0 {
+                        // This is the last border, so we need to include it
+                        max_estimate += 1;
+                    }
```
Comment (coszio) [inline lib/segment/src/index/field_index/histogram.rs:345]: This change is not related to this PR now that we are using the btree stream approach, but I'm pretty sure it is a correct fix... _maybe_?

File: lib/collection/src/collection_manager/segments_searcher.rs [hunk 9]
```diff
@@ -332,13 +332,21 @@ impl SegmentsSearcher {
...
                 point_version.insert(id, version);
             }
             Ok(true)
         })?;
-        Ok(point_records.into_values().collect())
+
+        // Restore the order the ids came in
+        let ordered_records = points
+            .iter()
+            .flat_map(|point| point_records.remove(point))
```
[the remark this answers is not in the document]
Comment (coszio) [reply lib/collection/src/collection_manager/segments_searcher.rs:345]: But it will take more memory. No concerns there?

File: lib/collection/src/operations/types.rs [hunk 10]
```diff
@@ -90,6 +92,8 @@ pub struct Record {
     /// Shard Key
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub shard_key: Option<ShardKey>,
+    #[serde(skip)]
+    pub ordered_by: Option<OrderedFloat<f64>>,
```
[the remark this answers is not in the document]
Comment (coszio) [reply lib/collection/src/operations/types.rs]: Fixed this by including this value in the payload for crossing the remote shard boundary

File: lib/segment/src/types.rs (hunk 3 above)
Comment (coszio) [reply lib/segment/src/types.rs:1093]: Fixed 👍

File: lib/segment/src/index/field_index/histogram.rs (hunk 4 above)
Comment (coszio) [reply lib/segment/src/index/field_index/histogram.rs]: Not needed anymore

File: lib/segment/src/index/field_index/histogram.rs (hunk 5 above)
Comment (coszio) [reply lib/segment/src/index/field_index/histogram.rs]: Not needed anymore

**Comment by coszio** [comment]
Update: 

Re-implemented most of the PR to follow a different approach. Now it uses a stream of value-ordered points which comes from the numeric index's `BTree`. This avoids fetching extra points that the range estimation approach could cause.

re: @generall 
> Also, I would like a bit more detailed comment on what happens if:
> 
> 1. offset is removed from the storage in between scroll calls
> 2. there are multiple numeric values per point

1. It would behave the same as not requesting an offset. This can be easily fixed by exposing `order_by.offset_value` and letting the user protect against this case themselves, but this is more of a design choice. What would you prefer?
2. They should appear twice in the search, regardless of being on the same page or not. I'll add a test to validate this.

**Comment by coszio** [comment]
Fixed multi-value case: Now each point is unique for every (value, id) combination that comes from the order_by field.

It is now ready for another review 
cc @general

File: lib/collection/src/collection_manager/segments_searcher.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (coszio) [reply lib/collection/src/collection_manager/segments_searcher.rs]: Actually this was deduplicating the input. Switched to getting the points in the same order as they came in, and letting the hashmap be dropped with the function block

File: lib/segment/src/segment.rs [hunk 11]
```diff
@@ -618,6 +623,19 @@ impl Segment {
             .collect()
     }
 
+    pub fn read_by_id_stream(
```
Comment (generall) [inline lib/segment/src/segment.rs:626]: Is this function used?

File: lib/segment/src/segment.rs [hunk 12]
```diff
@@ -666,6 +686,56 @@ impl Segment {
...
+        let id_tracker = self.id_tracker.borrow();
+
+        let range_iter = numeric_index.stream_with_value(&order_by.as_range());
+
+        let ordered_iter =
+            OrderedByFieldIterator::new(range_iter, id_tracker.deref(), order_by.direction())
+                .skip_while(|item| match offset {
+                    Some(id_offset) => {
+                        let item_key = (item.value, item.external_id);
+                        let offset_key = (order_by.value_offset(), id_offset);
```
Comment (generall) [inline lib/segment/src/segment.rs:713]: this part can be pre-computer outside of the loop

File: lib/segment/src/data_types/order_by.rs [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+            Direction::Desc => Range {
+                lte: Some(from),
+                ..Default::default()
+            },
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Validate, Clone, Default)]
+#[serde(deny_unknown_fields)]
```
Comment (generall) [inline lib/segment/src/data_types/order_by.rs:36]: Would it be possible to avoid that?
One possible pitfal here is that we require offset point to be not deleted between iterations, which might sometimes be problematic.
So maybe instead we could allow user to define `value_offset` explicitly?

Comment (coszio) [reply lib/segment/src/data_types/order_by.rs:36]: Sure, it was something that couldn't convince myself, but exposing it makes a lot of sense

File: lib/segment/src/segment.rs (hunk 11 above)
Comment (coszio) [reply lib/segment/src/segment.rs:626]: Thanks for noticing, I meant to extract it from `filtered_read()`

File: lib/segment/src/segment.rs (hunk 12 above)
Comment (coszio) [reply lib/segment/src/segment.rs:713]: Yeah, that's right :D

File: lib/segment/src/data_types/order_by.rs (hunk 13 above)
Comment (coszio) [reply lib/segment/src/data_types/order_by.rs]: Done

**Comment by coszio** [comment]
(sorry for the rebase, my local setup was having cache issues and thought it needed a rebase for some reason)

**Comment by coszio** [comment]
This PR has become too noisy, and the approach has changed enough, so I'll close it in favor of the new and cleaner one #3493
